### PR TITLE
[Serializer] Fix ignored callbacks in denormalization

### LINKED ISSUE #43563: [Serializer > Context (annotation) > Callbacks] Callbacks defined in the attribute Context annotation are ignored
**Symfony version(s) affected**: 5.3

**Description**
[According to this post](https://symfony.com/blog/new-in-symfony-5-3-inlined-serialization-context), the Context annotation allows to add context per attribute.

If you look closely into the code though, you'll discover that

1. Upon normalization, the callbacks are only retrieved from the context passed explicitly to ```normalize()``` method:
https://github.com/symfony/serializer/blob/5d7f068253ac3e7c62964ebdda491b06d401059a/Normalizer/AbstractObjectNormalizer.php#L187

2. Upon denormalization, either callbacks passed explicitly to ```denormalize()``` or passed as annotation, e.g.
```php

/**
 * @Context({
 *     "denormalizationContext": { "callbacks": { "someValue": "doSomeStuffOnDenormalization" } }
 * })
 */
private int $someValue;
```
are completely ignored. (?)

https://github.com/symfony/serializer/blob/5d7f068253ac3e7c62964ebdda491b06d401059a/Normalizer/AbstractObjectNormalizer.php#L401

**How to reproduce**
Add Context annotation to an attribute, with callbacks for normalization / denormalization and see if they are actually called.

**Possible Solution**
Add the missing functionality

### PR DESCRIPTION
**Pull request by benjaminmal**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #43563-2
| License       | MIT
| Doc PR        | -

Stop ignoring callbacks in denormalization. #freeTheCallbacks!!!

### CODE REVIEW

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php [hunk 1]
```diff
@@ -565,4 +555,56 @@ protected function createChildContext(array $parentContext, $attribute/*, ?strin
...
+
+    /**
+     * Apply callbacks set in context.
+     *
+     * @param mixed         $value
+     * @param string|object $object Can be either the object being normalizing or the object denormalized class's
+     *
+     * @return mixed
+     */
+    protected function applyCallbacks($value, $object, string $attribute, ?string $format, array $context)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: let's make both new methods `final`

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php [hunk 2]
```diff
@@ -565,4 +555,56 @@ protected function createChildContext(array $parentContext, $attribute/*, ?strin
...
+     *
+     * @return void
+     *
+     * @throws InvalidArgumentException
+     */
+    protected function validateCallbackContext(array $context, string $contextType = '')
+    {
+        $contextType = '' !== $contextType ? " $contextType" : '';
+
+        if (isset($context[self::CALLBACKS])) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: this should be move first in the method, as a guard clause

File: src/Symfony/Component/Serializer/Tests/Normalizer/Features/CallbacksTestTrait.php [hunk 3]
```diff
@@ -28,6 +33,85 @@ public function testCallbacks($callbacks, $valueBar, $result)
...
+     * @dataProvider provideNormalizeCallbacks
+     */
+    public function testNormalizeCallbacksWithTypedProperty($callbacks, $valueBar, $result)
+    {
+        $normalizer = $this->getNormalizerForCallbacksWithPropertyTypeExtractor();
+
+        $obj = new CallbacksObject();
+        $obj->bar = $valueBar;
+
+        $this->assertEquals(
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Tests/Normalizer/Features/CallbacksTestTrait.php]: on one line
if we can use asserSame, let's use it instead (maybe we can't, I didn't check)

File: src/Symfony/Component/Serializer/Tests/Normalizer/Features/CallbacksTestTrait.php [hunk 4]
```diff
@@ -28,6 +33,85 @@ public function testCallbacks($callbacks, $valueBar, $result)
...
+
+        $obj = new class() extends CallbacksObject {
+            public function __construct()
+            {
+            }
+        };
+
+        $obj->bar = $valueBar;
+
+        $this->assertEquals(
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Tests/Normalizer/Features/CallbacksTestTrait.php]: same comment

**Review by dunglas — APPROVED** [review]
I'm in favor of deprecating the callbacks and recommending using custom normalizers instead and I would revert the changes made to the PHPDoc for consistency, otherwise +1.

**Comment by nicolas-grekas** [comment]
Thank you @benjaminmal.

**Comment by nicolas-grekas** [comment]
I just found that this is already deprecated and removed from 5.4.

**Comment by nicolas-grekas** [comment]
Hum, only `setCallbacks()` is deprecated, my bad.

**Pull request merged**